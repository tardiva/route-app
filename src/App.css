.app__sidebar {
  padding: 15px;
  background-color: #fff;
  width: 85%;
  max-width: 330px;
}

.app__sidebar-top {
  margin-bottom: 30px;
}