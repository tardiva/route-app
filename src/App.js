import React, { Component } from 'react';
import './App.css';
import Form from './Components/Form/Form';
import List from './Components/List/List';
import MapWrapper from './Components/Map/MapWrapper';
import Sidebar from 'react-sidebar';
import ToggleButton from './Components/ToggleButton/ToggleButton';
import update from 'immutability-helper';

const mql = window.matchMedia('(min-width: 768px)');

class App extends Component {

  constructor(props) {
    super(props);
    this.state = {
      scriptError: false,
      points: [],
      currentCoord: {lat: 55.7526443, lng: 37.62370929999997},
      currentId: 0,
      sidebarOpen: true,
      sidebarDocked: mql.matches
    };
    this.toggleSidebar = this.toggleSidebar.bind(this);
    this.toggleSidebar = this.toggleSidebar.bind(this);
    this.handleMediaQueryChanged = this.handleMediaQueryChanged.bind(this);
    this.addPoint = this.addPoint.bind(this);
    this.removePoint = this.removePoint.bind(this);
    this.reorderPoints = this.reorderPoints.bind(this);
    this.updateCurrentCoord = this.updateCurrentCoord.bind(this);
    this.movePoint = this.movePoint.bind(this);
    this.callAfterScriptLoads = this.callAfterScriptLoads.bind(this);
  }

  componentWillMount() {
    mql.addListener(this.handleMediaQueryChanged);
  }

  render() {
    const sidebar =
      <div>
        <div className="app__sidebar-top">
          <Form onSubmit={this.addPoint}/>
        </div>
        <List items={this.state.points}
              moveItem={this.reorderPoints}
              removeItem={this.removePoint}/>
      </div>;

    const mainContent =
      <div>
        <MapWrapper warning={this.state.scriptError}
                    center={this.state.currentCoord}
                    points={this.state.points}
                    handleMapChange={this.updateCurrentCoord}
                    handlePointMove={this.movePoint}
                    asyncScriptOnLoad={this.callAfterScriptLoads}
        />
        {!this.state.sidebarDocked ? (<ToggleButton handleClick={this.toggleSidebar} />) : ''}
      </div>;

    return (
      <div className="App">
        <Sidebar
          open={this.state.sidebarOpen}
          docked={this.state.sidebarDocked}
          sidebarClassName="app__sidebar"
          onSetOpen={this.toggleSidebar}
          sidebar={sidebar}
          children={mainContent}
        >
        </Sidebar>
      </div>
    );
  }

  callAfterScriptLoads(resp) {
    this.setState(
      update(this.state, {
        scriptError: {
          $set: resp.errored
        }
      })
    )
  }

  toggleSidebar() {
    this.setState(
      update(this.state, {
        sidebarOpen: {
          $set: !this.state.sidebarOpen
        }
      })
    )
  }

  handleMediaQueryChanged() {
    this.setState(update(this.state, {
      sidebarDocked: {
        $set: mql.matches
      },
      sidebarOpen: {
        $set: false
      }
    }));
  }

  updateCurrentCoord(coord) {
    this.setState(
      update(this.state, {
        currentCoord: {
          $set: coord
        }
      })
    )
  }

  addPoint(name) {
    this.setState(update(this.state, {
      points: {
        $push: [{id: this.state.currentId, name: name, coord: this.state.currentCoord}]
      },
      currentId: {
        $set: this.state.currentId + 1
      }
    }));
  }

  removePoint(index) {
    this.setState(
      update(this.state, {
        points: {
          $splice: [[index, 1]]
        }
      })
    )
  }

  reorderPoints(dragIndex, hoverIndex) {
    this.setState(
      update(this.state, {
        points: {
          $splice: [[dragIndex, 1], [hoverIndex, 0, this.state.points[dragIndex]]]
        }
      })
    )
  }

  movePoint(index, coord) {
    this.setState(
      update(this.state, {
        points: {
          [index]: {
            coord: {
              $set: coord
            }
          }
        }
      })
    )
  }
}

export default App;
