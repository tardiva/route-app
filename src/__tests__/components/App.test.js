import React from 'react';
import { shallow } from 'enzyme';
import '../../__mocks__/matchMedia.mock';
import App from '../../App';
import points from '../../Fixtures/points';

it('should render App correctly', () => {
  const wrapper = shallow(<App/>);
  expect(wrapper).toMatchSnapshot();
});

it('should update state wnen new point was added', () => {
  const wrapper = shallow(<App/>);
  const component = wrapper.instance();
  component.setState({points: []});
  const point = {id: component.state.currentId, name: 'My New point', coord: component.state.currentCoord};
  component.addPoint(point.name);
  expect(component.state.points.length).toEqual(1);
  expect(component.state.points[0]).toEqual(point);
});

it('should update state wnen new point was deleted', () => {
  const wrapper = shallow(<App/>);
  const component = wrapper.instance();
  component.setState({points: points});
  const point = points[0];
  component.removePoint(point.id);
  expect(component.state.points.length).toEqual(points.length - 1);
  expect(component.state.points).not.toEqual(expect.arrayContaining([points[0]]));
});

it('should update state wnen reordering points', () => {
  const wrapper = shallow(<App/>);
  const component = wrapper.instance();
  component.setState({points: points});
  component.reorderPoints(1, 0);
  expect(component.state.points.length).toEqual(points.length);
  expect(component.state.points[0]).toEqual(points[1]);
  expect(component.state.points[1]).toEqual(points[0]);
});

it('should update state wnen current coordinates change', () => {
  const wrapper = shallow(<App/>);
  const component = wrapper.instance();
  component.setState({currentCoord: {lat: 0, lng: 0}});
  const coord = {lat: 1, lng: 1};
  component.updateCurrentCoord(coord);
  expect(component.state.currentCoord).toEqual(coord);
});

it('should update state wnen point coordinates change', () => {
  const wrapper = shallow(<App/>);
  const component = wrapper.instance();
  component.setState({points: points});
  const index = 0;
  const coord = {lat: 5, lng: 5};
  component.movePoint(0, coord);
  expect(component.state.points[index].coord).toEqual(coord);
});

it('should update state wnen resizing window', () => {
  const wrapper = shallow(<App/>);
  const component = wrapper.instance();
  component.setState({sidebarOpen: true, sidebarDocked: true});
  component.handleMediaQueryChanged();
  expect(component.state.sidebarOpen).toEqual(false);
  expect(component.state.sidebarDocked).toEqual(false);
});

it('should update state wnen toggling sidebar', () => {
  const wrapper = shallow(<App/>);
  const component = wrapper.instance();
  const isOpen = true;
  component.setState({sidebarOpen: isOpen});
  component.toggleSidebar();
  expect(component.state.sidebarOpen).toEqual(!isOpen);
});

it('should update state wnen script loading error occurs', () => {
  const wrapper = shallow(<App/>);
  const component = wrapper.instance();
  const response = {errored: true};
  component.callAfterScriptLoads(response);
  expect(component.state.scriptError).toEqual(response.errored);
});






