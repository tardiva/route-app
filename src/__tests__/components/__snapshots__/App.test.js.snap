// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render App correctly 1`] = `
<div
  className="App"
>
  <Sidebar
    defaultSidebarWidth={0}
    docked={false}
    dragToggleDistance={30}
    onSetOpen={[Function]}
    open={true}
    pullRight={false}
    shadow={true}
    sidebar={
      <div>
        <div
          className="app__sidebar-top"
        >
          <CreateItemForm
            onSubmit={[Function]}
          />
        </div>
        <DragDropContext(List)
          items={Array []}
          moveItem={[Function]}
          removeItem={[Function]}
        />
      </div>
    }
    sidebarClassName="app__sidebar"
    styles={Object {}}
    touch={true}
    touchHandleWidth={20}
    transitions={true}
  >
    <div>
      <AsyncScriptLoader(Map)
        asyncScriptOnLoad={[Function]}
        center={
          Object {
            "lat": 55.7526443,
            "lng": 37.62370929999997,
          }
        }
        handleMapChange={[Function]}
        handlePointMove={[Function]}
        points={Array []}
        warning={false}
      />
      <ToggleButton
        handleClick={[Function]}
      />
    </div>
  </Sidebar>
</div>
`;
